include/boost/geometry/util/type_traits.hpp

```diff
@@ -0,0 +1,323 @@
+namespace boost { namespace geometry
+{
+
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail
+{
+
+
+// C++17
+template <bool B>
+using bool_constant = std::integral_constant<bool, B>;
+
+
+// C++17
+template <typename ...>
+struct conjunction
+    : std::true_type
+{};
+template<typename Trait>
+struct conjunction<Trait>
+    : Trait
+{};
+template <typename Trait, typename ...Traits>
+struct conjunction<Trait, Traits...>
+    : std::conditional_t<Trait::value, conjunction<Traits...>, Trait>
+{};
+
+
+// C++17
+template <typename ...>
+struct disjunction
+    : std::false_type
+{};
+template <typename Trait>
+struct disjunction<Trait>
+    : Trait
+{};
+template <typename Trait, typename ...Traits>
+struct disjunction<Trait, Traits...>
+    : std::conditional_t<Trait::value, Trait, disjunction<Traits...>>
+{};
+
+
+// C++17
+template <typename Trait>
+struct negation
+    : bool_constant<!Trait::value>
+{};
+
+
+/*
+template <typename ...Traits>
+using and_ = conjunction<Traits...>;
+
+template <typename ...Traits>
+using or_ = disjunction<Traits...>;
+
+template <typename Trait>
+using not_ = negation<Trait>;
+*/
+
+
+template <typename T>
+struct is_geometry
+    : bool_constant<! std::is_void<typename tag<T>::type>::value>
+{};
+
+template <typename T>
+struct is_not_geometry
+    : std::is_void<typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_point
+    : std::is_same<point_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_multi_point
+    : std::is_same<multi_point_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_pointlike
+    : std::is_base_of<pointlike_tag, typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_segment
+    : std::is_same<segment_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_linestring
+    : std::is_same<linestring_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_multi_linestring
+    : std::is_same<multi_linestring_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_polylinear
+    : std::is_base_of<polylinear_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_linear
+    : std::is_base_of<linear_tag, typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_box
+    : std::is_same<box_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_ring
+    : std::is_same<ring_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_polygon
+    : std::is_same<polygon_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_multi_polygon
+    : std::is_same<multi_polygon_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_polygonal
+    : std::is_base_of<polygonal_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_areal
+    : std::is_base_of<areal_tag, typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_polysegmental
+    : bool_constant<is_polylinear<T>::value || is_polygonal<T>::value>
+{};
```

Comment:
Here is another group of geometries, i.e. geometries based on ranges of segments which are not a single segment, both linear and areal. I named it `polysegmental` feel free to suggest a different name.